### PR TITLE
[RFC] Support custom waiting logic for query in coordinator

### PR DESCRIPTION
**Pull request by mayankgarg1990**
Support custom logic to make the coordinator wait for business logic to finish before the query is ready to run. This is to support situations where in different implementations might want for some data to be generated or replicated before the query actually starts.

This PR is a work in progress and is created just to gather feedback on the design. The high level approach is the following:

1. Have a new state `WAITING_FOR_PREREQUISITES` which indicates that the query is waiting for some logic to finish before it is queued
2. Introduce a SPI interface `PrerequisiteManager` which lets different Presto deployments implement this optional logic if they want to use this functionality

```
== RELEASE NOTES ==

General Changes
* TBD

SPI Changes
* TBD
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/server/CoordinatorModule.java [hunk 1]
```diff
@@ -251,6 +253,7 @@ protected void setup(Binder binder)
         binder.bind(DispatchManager.class).in(Scopes.SINGLETON);
         binder.bind(FailedDispatchQueryFactory.class).in(Scopes.SINGLETON);
         binder.bind(DispatchExecutor.class).in(Scopes.SINGLETON);
+        newOptionalBinder(binder, PrerequisiteManager.class);
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/server/CoordinatorModule.java:256]: A better way to model this is through a Plugin.  i.e., one should just be able to specify a Plugin, which may or may not be packaged with the core Presto distribution, to override this in a customizeable way.  See examples of `TempStorageFactory`, `FunctionNamespaceManagerFactory`, etc. in `PluginManager`.

**Comment by xumingming** [comment]
@mayankgarg1990 can you share your actual scenario why you need this waiting phase?

**Comment by mayankgarg1990** [comment]
@xumingming - we want support running queries across 2 different datacenters - but depending on the size of the data, we may decide to copy the data to prevent runtime join on remote data. So the wait will give us an opportunity to copy data over

**Comment by mayankgarg1990** [comment]
Closed in favor of https://github.com/prestodb/presto/pull/16073

**Pull request closed without merging**